### ISSUE TITLE
connected peers appears to contain duplicates

### ISSUE THREAD

**Issue by antiochp** [mimblewimble/grin#228]
See #225 for the new peers endpoints.
The connected peers endpoint pulls data from the p2p_server list of peers.

Leaving a couple of nodes running for a few minutes results in the following (one node has the other node listed as a connected peer multiple times) -

```
curl 127.0.0.1:13413/v1/peers/connected
[
  {
    "capabilities": {
      "bits": 7
    },
    "user_agent": "MW/Grin 0.1",
    "version": 1,
    "addr": "127.0.0.1:14414",
    "total_difficulty": 37
  },
  {
    "capabilities": {
      "bits": 7
    },
    "user_agent": "MW/Grin 0.1",
    "version": 1,
    "addr": "127.0.0.1:14414",
    "total_difficulty": 37
  },
  {
    "capabilities": {
      "bits": 7
    },
    "user_agent": "MW/Grin 0.1",
    "version": 1,
    "addr": "127.0.0.1:14414",
    "total_difficulty": 37
  },
  {
    "capabilities": {
      "bits": 7
    },
    "user_agent": "MW/Grin 0.1",
    "version": 1,
    "addr": "127.0.0.1:14414",
    "total_difficulty": 37
  },
  {
    "capabilities": {
      "bits": 7
    },
    "user_agent": "MW/Grin 0.1",
    "version": 1,
    "addr": "127.0.0.1:14414",
    "total_difficulty": 37
  }
]
```

This is hitting the endpoint on the node running on 13414.
14414 is the only peer it is connected to.
(both running locally, hence 127.0.0.1).

**Comment by antiochp** [comment]
If our p2p server listens on `0.0.0.0:13413`
And the other peers can resolve our ip address correctly via `extract_ip` on their end.
And then peers send over an updated peer list -
How do we know which one corresponds to our own node? We don't necessarily know our ip address, even though our peers do.

**Comment by ignopeverell** [comment]
Correct, so the handshake includes a randomly generated nonce to detect self-connects:

https://github.com/mimblewimble/grin/blob/master/p2p/src/handshake.rs#L38

**Comment by kargakis** [comment]
> How do we know which one corresponds to our own node? We don't necessarily know our ip address, even though our peers do.

There are cases where you know your IP either beforehand (eg. a static IP) or at runtime w/o shelling out to your host (eg. downward API in Kubernetes). Is there any downside to ignoring a blacklist before doing the handshake?

**Comment by antiochp** [comment]
Update on this since the fix went in handle `127.0.0.1` and `0.0.0.0` correctly - 

See below for peers info on a "network" of 3 nodes.
* 1 seed (points to no other seeds)
* 1 mining node (points to single seed node in seed list)
* 1 non-mining node (points to single seed node in seed list)

```
### seed node
curl 52.14.128.150:13413/v1/peers/connected
[{"capabilities":{"bits":7},"user_agent":"MW/Grin 0.1","version":1,"addr":"18.220.204.168:13414","total_difficulty":1},
{"capabilities":{"bits":7},"user_agent":"MW/Grin 0.1","version":1,"addr":"18.216.195.126:13414","total_difficulty":1}]%  


### mining node
curl 18.220.204.168:13413/v1/peers/connected
[{"capabilities":{"bits":7},"user_agent":"MW/Grin 0.1","version":1,"addr":"52.14.128.150:13414","total_difficulty":1},
{"capabilities":{"bits":7},"user_agent":"MW/Grin 0.1","version":1,"addr":"18.220.204.168:13414","total_difficulty":1},
{"capabilities":{"bits":7},"user_agent":"MW/Grin 0.1","version":1,"addr":"18.220.204.168:13414","total_difficulty":1},
{"capabilities":{"bits":7},"user_agent":"MW/Grin 0.1","version":1,"addr":"18.216.195.126:13414","total_difficulty":1}]%  


### non mining node
curl 18.216.195.126:13413/v1/peers/connected
[{"capabilities":{"bits":7},"user_agent":"MW/Grin 0.1","version":1,"addr":"52.14.128.150:13414","total_difficulty":7767},
{"capabilities":{"bits":7},"user_agent":"MW/Grin 0.1","version":1,"addr":"18.220.204.168:13414","total_difficulty":7783},
{"capabilities":{"bits":7},"user_agent":"MW/Grin 0.1","version":1,"addr":"18.216.195.126:13414","total_difficulty":7783},
{"capabilities":{"bits":7},"user_agent":"MW/Grin 0.1","version":1,"addr":"18.216.195.126:13414","total_difficulty":7783}]%   
```

So -
1) all nodes sees the other two nodes correctly, but -
2) the non-seed nodes appear to have _themselves_ in their peer list (twice)

**Comment by antiochp** [comment]
Duplicates should be resolved in #244.

**Comment by antiochp** [comment]
I think part of the problem here is related to `is_self()`.
This is not going to work as this is based on the host and port from config (and host here will be `0.0.0.0`)

https://github.com/mimblewimble/grin/blob/5b47da907f1ec16aa6c1c4d18697324ec89c0843/p2p/src/server.rs#L206-L208

This comes back to the earlier discussion around "how do we know that a given host and port is ourselves?"
The nonce in the handshake covers part of this but I'm not sure it covers everything here.

**Comment by antiochp** [comment]
I think part of the problem here is related to `is_self()`.
This is not going to work as this is based on the host and port from config (and host here will be `0.0.0.0`)

https://github.com/mimblewimble/grin/blob/5b47da907f1ec16aa6c1c4d18697324ec89c0843/p2p/src/server.rs#L206-L208

This comes back to the earlier discussion around "how do we know that a given host and port is ourselves?"
The nonce in the handshake covers part of this but I'm not sure it covers everything here.

**Comment by antiochp** [comment]
Both these issues should be resolved in #244 now.
